### PR TITLE
Generalized inter-group independencies

### PR DESCRIPTION
**Pull request by nnistelrooij**
Before, we had the `independent` parameter, which either kept or removed all the inter-group weights.
Instead of these two extremes, I've decided to generalize this method with Dropout.
If you set `dropout = 1.0`, then all the inter-group weights are dropped. So this setting is equivalent to `independent = True`.
If you set `dropout = 0.0`, then all the inter-group weights are kept. So this setting is equivalent to `independent = False`.

Hopefully, by setting `dropout` to a value somewhere between 0.0 and 1.0, we get the best of both worlds. (1) we make use of the weights that are not used when `independent = True` and (2) the model does not overfit, because we can tightly control its complexity.

This might just be the trick to get faster convergence for the LSTM method, while hopefully achieving better results than `Prophet`.

### CODE REVIEW

**Comment by nnistelrooij** [comment]
Okay, so this didn't work.

Setting the weights to zero, means that the gradient will be uninformative. For example: if the current weight is at 3.0, the optimal weight is at 2.5, and we set the weight to 0.0, then the gradient will be positive (away from 2.5), whereas it should be negative (toward 2.5). This can be fixed by only updating the weights that are not zeroed, but then each weight is only updated 1% of the time.

So, instead of that, I just decrease the learning rate of the inter-group weights by multiplying their gradients with a small factor in the backward hooks. So the inter-group weights are updated each iteration with an informative gradient, but they just learn a lot slower compared to the intra-group weights.

**Review by rusane — APPROVED** [review]
Nice work! The `DropOut` solution from before seems quite complex in terms of code compared to the learning rate for inter-group weights😅. When you set the learning rate for the inter-group weights to 0, then the model is the same as `inpdendent=True` in the old code right?

**Comment by nnistelrooij** [comment]
Not entirely, in the old code, the inter-group **weights** and gradients would be set to 0, whereas with this new approach, only the inter-group gradients are set to 0 when `inter_group_lr = 0.0`.

As you can see in the plot on WhatsApp, the model starts learning slower, because it has to adjust the biases to compensate for the non-zero inter-group weights. After it has done that, it picks up the pace.

When `inter_group_lr = 1.0`, then it is exactly the same as with `independent = False`; we don't touch the weights or the gradients.

File: main.py [hunk 1]
```diff
@@ -23,10 +23,10 @@ def handle_arguments():
                         help='number of days for validation, default: 28')
     parser.add_argument('-s', '--num_models', type=int, default=500,
                         help='number of submodels to make, default: 500')
-    parser.add_argument('-d', '--dependent', action='store_true',
-                        help='switch to have inter-group dependencies')
-    parser.add_argument('-e', '--epochs', type=int, default=50, help='number '
-                        'of iterations over training data, default: 50')
+    parser.add_argument('-l', '--inter_group_lr', type=float, default=0.0,
```
Comment (MauiGonzo) [inline main.py]: (no dependent choice anymore indeed)

File: main.py [hunk 2]
```diff
@@ -23,10 +23,10 @@ def handle_arguments():
...
     parser.add_argument('-s', '--num_models', type=int, default=500,
                         help='number of submodels to make, default: 500')
-    parser.add_argument('-d', '--dependent', action='store_true',
-                        help='switch to have inter-group dependencies')
-    parser.add_argument('-e', '--epochs', type=int, default=50, help='number '
-                        'of iterations over training data, default: 50')
+    parser.add_argument('-d', '--dropout', type=float, default=1.0,
+                        help='prob to drop inter-group weight grad, default: 1')
+    parser.add_argument('-e', '--epochs', type=int, default=50,
+                        help='number of iters over training data, default: 50')
```
Comment (MauiGonzo) [inline main.py:29]: isn't it confusing to set epochs to control iters

**Review by MauiGonzo — APPROVED** [review]

File: main.py (hunk 2 above)
Comment (nnistelrooij) [reply main.py:29]: It had to fit within 80 columns. It is probably better with 'iterations over training data, default: 50'
